### PR TITLE
Fix CPR bugs associated with non-unique timestamps by publish date

### LINKED ISSUE #1556: DSEW-CPR testing signals fail for Community_Profile_Report_20220225
Got the following output while trying out the new testing signals on staging:

```
[...]
{"filename": "./input_cache/Community Profile Report 20220225--539a654c-7b48-4730-b4c2-34f1c64b5d49.xlsx", "event": "Downloading file", "logger": "delphi_dsew_community_profile.run", "level": "info", "timestamp": "2022-03-11T15:42:11.893222Z"}
{"sheet": "Regions", "filename": "./input_cache/Community Profile Report 20220225--539a654c-7b48-4730-b4c2-34f1c64b5d49.xlsx", "event": "Building dfs", "logger": "delphi_dsew_community_profile.run", "level": "info", "timestamp": "2022-03-11T15:42:12.857682Z"}
{"sheet": "States", "filename": "./input_cache/Community Profile Report 20220225--539a654c-7b48-4730-b4c2-34f1c64b5d49.xlsx", "event": "Building dfs", "logger": "delphi_dsew_community_profile.run", "level": "info", "timestamp": "2022-03-11T15:42:12.972835Z"}
{"sheet": "CBSAs", "filename": "./input_cache/Community Profile Report 20220225--539a654c-7b48-4730-b4c2-34f1c64b5d49.xlsx", "event": "Building dfs", "logger": "delphi_dsew_community_profile.run", "level": "info", "timestamp": "2022-03-11T15:42:13.219382Z"}
{"sheet": "Counties", "filename": "./input_cache/Community Profile Report 20220225--539a654c-7b48-4730-b4c2-34f1c64b5d49.xlsx", "event": "Building dfs", "logger": "delphi_dsew_community_profile.run", "level": "info", "timestamp": "2022-03-11T15:42:15.193136Z"}
{"event": "Top-level exception occurred", "logger": "delphi_dsew_community_profile.run", "level": "error", "timestamp": "2022-03-11T15:42:23.090096Z", "exception":[reformatted below]}
Traceback (most recent call last):
  File "/home/indicators/.pyenv/versions/3.8.2/lib/python3.8/runpy.py", line 193, in _run_module_as_main
    return _run_code(code, main_globals, None,
  File "/home/indicators/.pyenv/versions/3.8.2/lib/python3.8/runpy.py", line 86, in _run_code
    exec(code, run_globals)
  File "/mnt/data/indicators/runtime/dsew_community_profile/delphi_dsew_community_profile/__main__.py", line 12, in <module>
    run_module(read_params())  # pragma: no cover
  File "/mnt/data/indicators/runtime/dsew_community_profile/delphi_dsew_community_profile/run.py", line 60, in run_module
    dfs = fetch_new_reports(params, logger)
  File "/mnt/data/indicators/runtime/dsew_community_profile/delphi_dsew_community_profile/pull.py", line 529, in fetch_new_reports
    ret[key] = unify_testing_sigs(
  File "/mnt/data/indicators/runtime/dsew_community_profile/delphi_dsew_community_profile/pull.py", line 603, in unify_testing_sigs
    volume_df = add_max_ts_col(volume_df)[
  File "/mnt/data/indicators/runtime/dsew_community_profile/delphi_dsew_community_profile/pull.py", line 641, in add_max_ts_col
    assert all(
AssertionError: Testing signals should have two unique timestamps per publish date-region combination.
```

A quick look at the excel file (attached) suggests a logic error somewhere, since the report does include both LAST WEEK and PREVIOUS WEEK for testing signals.

[Community_Profile_Report_20220225.xlsx](https://github.com/cmu-delphi/covidcast-indicators/files/8234901/Community_Profile_Report_20220225.xlsx)

### PR DESCRIPTION
**Pull request by nmdefries**
### Description
Fix two CPR bugs:

- Some asserts are too strict given earlier logic.
- On 2021-03-17, the reference dates for test volume and test positivity changed from being the same within a report to being different. This caused some timestamps to appear for multiple report dates, which the existing logic can't handle appropriately. Fix by splitting data into before/after that date and processing separately.

### Changelog
- `pull.py`
- tests

### Fixes 
Closes #1556.

### CODE REVIEW

File: dsew_community_profile/delphi_dsew_community_profile/pull.py
Comment (krivard) [inline dsew_community_profile/delphi_dsew_community_profile/pull.py:527]: Is this the same or different from `not latest_key_df.empty`?

File: dsew_community_profile/delphi_dsew_community_profile/pull.py
Comment (krivard) [inline dsew_community_profile/delphi_dsew_community_profile/pull.py:629]:
```suggestion
    # check that we have positivity *and* volume for each publishdate+geo
    pos_groups_df = positivity_df.drop(
```

File: dsew_community_profile/delphi_dsew_community_profile/pull.py
Comment (krivard) [inline dsew_community_profile/delphi_dsew_community_profile/pull.py:647]: Community_Profile_Report_20220225.xlsx has 158 counties where positivity is not available for "last week" but volume is, and 7 counties where volume is not available but positivity is. Will this assert inappropriately reject that file? Or do you mean that previous steps should have already dropped the entries with mismatched availability?

File: dsew_community_profile/delphi_dsew_community_profile/pull.py
Comment (krivard) [inline dsew_community_profile/delphi_dsew_community_profile/pull.py:688]: does this change mean that you found files that only have one week of test data in them, instead of both "last week" and "previous week"?

**Review by krivard — COMMENTED** [review]
Some missing documentation, and I may also be getting confused about one context of two reference dates ("last week" "previous week") and another ("last week" positivity vs volume after 2021-03-17)

File: dsew_community_profile/delphi_dsew_community_profile/pull.py
Comment (nmdefries) [reply dsew_community_profile/delphi_dsew_community_profile/pull.py:689]: In short, no, this is the bit that actually fixes issue #1556.

Previously, we expected that "[t]esting signals should have two unique timestamps per publish date-region combination." This is true of _raw_ data, but causes false positives for our processed data.

This particular check [is called in `unify_testing_sigs`](https://github.com/cmu-delphi/covidcast-indicators/blob/eaeda1d87b644726cc2f997cbb80db3869653732/dsew_community_profile/delphi_dsew_community_profile/pull.py#L558-L559) after we [deduplicate timestamps](https://github.com/cmu-delphi/covidcast-indicators/blob/eaeda1d87b644726cc2f997cbb80db3869653732/dsew_community_profile/delphi_dsew_community_profile/pull.py#L518-L525). We only keep those with the most recent publish date, which means that some publish date-timestamp combinations are dropped.

Specifically, if we're generating output over a wide enough time period so that we see a given date reported for "last week" and also for the "previous week", we will always keep the version reported for the "previous week" since it's publish date will be 7 days newer. As a result, the older publish date will have all of it's "last week" data discarded; it will only be associated with the "previous week"'s timestamp.

Comment (krivard) [reply dsew_community_profile/delphi_dsew_community_profile/pull.py:689]: Are there ever cases where only one timestamp is available for volume but two are available for positivity, or vice-versa?

File: dsew_community_profile/delphi_dsew_community_profile/pull.py
Comment (nmdefries) [reply dsew_community_profile/delphi_dsew_community_profile/pull.py:648]: The wording may be ambiguous on my part. "available" here means that there is a row for a given region and date. It doesn't check if the signal is `NA`or not. `pandas.read_excel` reads empty fields as `NA/NaN`, so obs are included as `NA` where the signal was not available in the original report.

The way the sample size filter is set up at the moment, any obs with missing (`NA`) n or n <= 5 are dropped.

Comment (nmdefries) [reply dsew_community_profile/delphi_dsew_community_profile/pull.py:648]: Rows are included if `val` (test positivity) is missing but sample size is available and > 5.

File: dsew_community_profile/delphi_dsew_community_profile/pull.py
Comment (nmdefries) [reply dsew_community_profile/delphi_dsew_community_profile/pull.py:689]: * checks that each publish date has the same geos for test volume and test positivity [here](https://github.com/cmu-delphi/covidcast-indicators/blob/ndefries/cpr-lenient-check-ts-per-publishdate/dsew_community_profile/delphi_dsew_community_profile/pull.py#L646-L648)
* checks that test volume and positivity have the same number of observations. Since we've already checked that they have the same geos by publish date, this ~checks that they have the same number of ts per publish date-geo.

Comment (nmdefries) [reply dsew_community_profile/delphi_dsew_community_profile/pull.py:689]: Assuming there are always the same number of timestamps available for test volume and positivity in the raw data, there should always be the same number in the processed data. The only logic that influences this is the timestamp deduplication (where we keep the most recent publish date for each timestamp). It is [applied to both test volume and positivity](https://github.com/cmu-delphi/covidcast-indicators/blob/ndefries/cpr-lenient-check-ts-per-publishdate/dsew_community_profile/delphi_dsew_community_profile/pull.py#L518-L523) using the same ["change date" threshold approach](https://github.com/cmu-delphi/covidcast-indicators/blob/4eee156c887eed6e25f1ddd0f9add27cd3213531/dsew_community_profile/delphi_dsew_community_profile/pull.py#L750).

This isn't strictly necessary for test positivity, but it makes it easier to match up volume and positivity signals for each publish date.

Empirically, running the pipeline over all reports with a new check for # of timestamps, I don't see any cases where the number of timestamps differs.

File: dsew_community_profile/delphi_dsew_community_profile/pull.py
Comment (krivard) [inline dsew_community_profile/delphi_dsew_community_profile/pull.py:783]: it looks like we never actually call this with different arguments to be used before vs after the change date. was that intended?

Comment (nmdefries) [reply dsew_community_profile/delphi_dsew_community_profile/pull.py:783]: Yeah, the before/after change date data is always processed using the same args.

The alternative implementation is to pass single values as `*args` instead of lists, and duplicate them to the right length within `apply_thres_change_date`.

**Review by krivard — APPROVED** [review]
lgtm! 👍  want me to wait for ananya before merging or no?

**Comment by Ananya-Joshi** [comment]
Tests and some sample dates I looked at from march 2022 seem to be working fine/generally in the right range. Happy to approve pending some information on the comments above. Good job Nat!

File: dsew_community_profile/delphi_dsew_community_profile/pull.py
Comment (nmdefries) [reply dsew_community_profile/delphi_dsew_community_profile/pull.py:723]: The first clause of the assert is checking that each publish date-geo combo should have up to 2 observations (AKA timestamps). The second clause is checking that when a publish date-geo combo has 2 observations/timestamps, there are also 2 _unique_ timestamps (of course, this is trivially true in the case that a publish date-geo combo only has 1 observation/timestamp).

Another way to check the second bit would be seeing if `all(assert_df.num_obs == assert_df.num_unique_obs)`. I can add that in if you find it clearer.

File: dsew_community_profile/delphi_dsew_community_profile/pull.py
Comment (nmdefries) [reply dsew_community_profile/delphi_dsew_community_profile/pull.py:770]: That's right, it started on [2021-03-17](https://github.com/cmu-delphi/covidcast-indicators/blob/3d1e60aadecc10415238959ad0cf4abedb71a3a2/dsew_community_profile/delphi_dsew_community_profile/pull.py#L790).

Comment (Ananya-Joshi) [reply dsew_community_profile/delphi_dsew_community_profile/pull.py:770]: Sorry, I meant there was a typo in reported and that it might be good to put that date in this comment!

File: dsew_community_profile/delphi_dsew_community_profile/pull.py
Comment (Ananya-Joshi) [reply dsew_community_profile/delphi_dsew_community_profile/pull.py:723]: Ah now I understand so it's all okay. I think your suggestion would make it clearer even in the case of 1 observation/timestamp.

**Review by Ananya-Joshi — APPROVED** [review]

**Pull request merged**
